Add multi-account support with custom entity prefixes

llaforest opened the pull request:
- Add 'accounts' configuration option to support multiple Neviweb accounts
- Each account can have a custom prefix for entity naming
- Maintain full backward compatibility with existing single-account configs
- Support both 'location' and 'network' parameter names
- Update all platform files (climate, switch, sensor, light, valve)
- Update README with multi-account configuration examples"

llaforest commented:
Please try it, I had trouble to test the backward compatibility with my setup, because I used to rename all the neviweb130 and copy the folders before, my entity ids are not being regenerated the same way out of the box. You should not have to change anything in the configuration.yaml.
Then you can try to change the configuration.yaml with new structure with one account, or more.

claudegel commented:
I'll look at it this weekend after an update of another branch.
But it is easy to have more then one account in neviweb130 without changing anything in the code. You just need to copy all file from the neviweb130 dir to a new dir named let say neviweb131. after in that new dir just edit manifest.json and change line two:
"domain": "neviweb130",  ---> "domain": "neviweb131",
You will have two neviweb130 running together with all device named like this:
climate.neviweb130_climate_my_thermostat
in the other it will be named:
climate.neviweb131_climate_my_other_thermostat
But I like your idea. We are just finishing a big update to neviweb130 to go to version 4.0.0 and after I can check to add your PR.
You can check my other branch neviweb130-V2 which is a config_flow version. If you try it it will migrate your device unique_id as requested by HA from int to str. So if you go back to your regular neviweb130, HA will change all your device name by adding a _2 at the end. So if you can. a complete backup or another dev setup to test .
Thanks for your PR and proposal to help. It should make neviweb130 better if many people help me.

llaforest commented:
I added you as collaborator bassdr in my fork. I definitely have time to do it this weekend but I'm overwhelmed at work this week.
I will try to check the commit you do to resolve conflicts.

llaforest commented:
@bassdr it's a nice work you did there. This 4.0 release is massive, your rebase seems to be fine and I see my code on top of it. Will include it in my fork and update the PR accordingly is that ok?

claudegel commented:
Maybe it's better to rebase on fix-v4 as I've added to master in that branch. Once it is all ok there I'll merge to master.

llaforest commented:
The work you did @bassdr is very nice. I tried to look at a rebase from scratch to see the magnitude of the conflicts. It is huge and I would have has a lot of trouble resolving it myself. I think I would have done my feature again from scratch.

Anyways, I fetched your changed and pushed them to the PR. I see you merged with master instead of fix-v4. Anyway thanks for your work!

Let me know if you need me at some point for something.

bassdr commented:
> @claudegel 
> Maybe it's better to rebase on fix-v4 as I've added to master in that branch. Once it is all ok there I'll merge to master.

You probably should do a PR for this fix-v4, I can checkout the branch manually to validate but I won't be able to review it if it's not a PR.

Also, I am not sure what you mean with "rebase on fix-v4", if you meant really "rebase" it would mean this PR would effectively include fix-v4 as well, it's a bit unusual to rebase on someone else's branch in a PR as it basically means you are including work you did not do yourself. I know I did it with th6xxxwf but I did make sure I understand everything you did before proposing a PR. Still quite unusual but made more sense since I own the device and only I could test the features you did.

If you meant to target fix-v4 in this PR, it makes more sense, but then it's the opposite, you will take ownership of what @llaforest did here when you merge fix-v4 to master later, and the merge to master will become much harder to review. Since you are the maintainer it's really your call :)

bassdr commented:
Oops, juste saw #412 was merged before I could notice it exists, my bad. Merging master again goes straight forward tho.

claudegel commented:
If you want to manage two different account on Neviweb. For now you should go the way it is describe in the readme.md.
You need to copy all files from config/custom_components/neviweb130 to a new directory like config/custom_components/neviweb131.  (131 instead of 130)
In that dir, edit the file manifest.json and change the line "domain": "neviweb130", to "domain": "neviweb131", 
Then in configuration.yaml add a second config like this:
First config,
```
neviweb130:
  username: !secret climate_username
  password: !secret climate_password
  network: !secret climate_gateway130
  scan_interval: 360
  stat_interval: 600
  notify: "notification"
```
and a second config,
```
neviweb131:
  username: !secret climate_username_131
  password: !secret climate_password_131
  network: !secret climate_gateway_131
  scan_interval: 360
  stat_interval: 600
  notify: "notification"
```
Make sure you put the wright credential for the second account on Neviweb.
Should work. See the readme.md

claudegel commented:
For multi site you need to have more then one network in your config. You can have up to 3 network, (location) on the same account. network, network2 and network3

claudegel commented:
I just push a new version v4.0.0 that include all late modifications to master. Now I'll have a look to this PR

claudegel commented:
Yes In Neviweb you can have two locations per account and when you ask you can get a third one. This is what I have as I needed one to test neviweb130-V2 (version v5.0.0). I can switch that third one into a second account so I'll get on account with two locations and another with one location.
If you can rebase on master I'll check and fix. I also will need to add this code in neviweb130-V2 to make sure we don't break anything.

claudegel commented:
I've almost finish adding two account support in neviweb130-V2. I've some questions. I've added a prefix to distinguish device from different account and I'm using the entry.entry_id to differentiate the unique_id of each device between each account.
Now for example climate device are named like climate.neviweb130_climate_TH1124WF
for main account no prefix is added. For other account a prefix is added as climate.neviweb130_climate_prefix_TH1124WF
Prefix is a new parameter with default = ""
I don't know if it is better to just replace the default name neviweb130_climate by neviweb130_prefix and for other location in same account neviweb130_prefix_2 and neviweb130_prefix_3. This would break all automation as they are based on name.
Or maybe instead of a prefix we can add a suffix like climate.neviweb130_climate_TH1124WF_suffix
What do you think ?

claudegel commented:
Yes I thing prefix is better. But we cannot remove _climate after neviweb130 (climate.neviweb130_climate_bathroom) or it will break all automations. But as we keep prefix = "" for the first account we are safe.
So I suppose that we can have: 
climate.neviweb130_climate_bathroom, for existing account
climate.neviweb130_parents_chalet_bathroom, where parent is account name and chalet is location name.
as we already have in the code for all platform:
DEFAULT_NAME = f"{DOMAIN} climate"
DEFAULT_NAME_2 = f"{DOMAIN} climate 2"
DEFAULT_NAME_3 = f"{DOMAIN} climate 3"
we can have default prefix = climate
default location 1 = ""
account 2 prefix = parents
account 2 location 1 = chalet
and then we can say DEFAULT_NAME = f"{DOMAIN} {prefix} {location}" 
first account will be DEFAULT_NAME = neviweb130_climate
second account will be DEFAULT_NAME = neviweb130_parents_chalet
Still need to patch for location 1 that is none presently but it is not a big deal.
If later the user want to change prefix for first account we can put a warning and documentation so he is aware of breaking his automations.

llaforest commented:
Let me know if you want me to make a new commit that does that tomorrow morning.

llaforest commented:
Ok I tested these scenarios:
In a Home Assistant with legacy config, updated to latest code, no entity issues, nothing changed.
In a Home Assistant  with new accounts: section, have three accounts with only location defined. all entities with neviweb130_location_device..... All good.
Added another account with two locations, and it appears nicely too.
